pkg/distributor/distributor_test.go

```diff
@@ -2616,3 +2620,216 @@ func TestDistributor_PushIngestLimits(t *testing.T) {
+		ingestLimitsEnabled:       true,
+		ingestLimitsDryRunEnabled: true,
+		tenant:                    "test",
+		streams: logproto.PushRequest{
+			Streams: []logproto.Stream{{
+				Labels: "{foo=\"bar\"}",
+				Entries: []logproto.Entry{{
+					Timestamp: time.Now(),
+					Line:      "baz",
+				}},
+			}},
+		},
+		limitsResponse: &logproto.ExceedsLimitsResponse{
+			Tenant: "test",
+			Results: []*logproto.ExceedsLimitsResult{{
+				StreamHash: 0x90eb45def17f924,
+				Reason:     limits_frontend.ReasonExceedsMaxStreams,
+			}},
+		},
+		// In dry-run mode, streams that are "dropped" should not be pushed
+		// to the metadata topic, otherwise dropped streams will be counted as
+		// existing streams.
+		expectedMetadataRecords: 0,
+	}, {
+		name:                "error checking limits",
+		ingestLimitsEnabled: true,
+		tenant:              "test",
+		streams: logproto.PushRequest{
+			Streams: []logproto.Stream{{
+				Labels: "{foo=\"bar\"}",
+				Entries: []logproto.Entry{{
+					Timestamp: time.Now(),
+					Line:      "baz",
+				}},
+			}},
+		},
+		limitsResponseErr:       errors.New("failed to check limits"),
+		expectedMetadataRecords: 1,
+	}, {
+		name:                      "error checking limits in dry-run mode",
+		ingestLimitsEnabled:       true,
+		ingestLimitsDryRunEnabled: true,
+		tenant:                    "test",
+		streams: logproto.PushRequest{
+			Streams: []logproto.Stream{{
+				Labels: "{foo=\"bar\"}",
+				Entries: []logproto.Entry{{
+					Timestamp: time.Now(),
+					Line:      "baz",
+				}},
+			}},
+		},
+		limitsResponseErr:       errors.New("failed to check limits"),
+		expectedMetadataRecords: 1,
+	}}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			limits := &validation.Limits{}
+			flagext.DefaultValues(limits)
+			distributors, _ := prepare(t, 1, 3, limits, nil)
+			d := distributors[0]
+			d.cfg.KafkaEnabled = true
+			d.cfg.IngesterEnabled = false
+			d.cfg.IngestLimitsEnabled = test.ingestLimitsEnabled
+			d.cfg.IngestLimitsDryRunEnabled = test.ingestLimitsDryRunEnabled
+			// The default value is 0, which means no records can be produced.
+			d.cfg.KafkaConfig.ProducerMaxRecordSizeBytes = 1000
+			mockKafkaClient := mockKafkaProducer{}
+			d.kafkaWriter = &mockKafkaClient
+			mockLimitsClient := mockIngestLimitsFrontendClient{
+				t:           t,
+				response:    test.limitsResponse,
+				responseErr: test.limitsResponseErr,
+			}
+			l := newIngestLimits(&mockLimitsClient, prometheus.NewRegistry())
+			d.ingestLimits = l
+
+			ctx = user.InjectOrgID(context.Background(), test.tenant)
+			resp, err := d.Push(ctx, &test.streams)
+			if test.expectedErr != "" {
+				require.EqualError(t, err, test.expectedErr)
+				require.Nil(t, resp)
+			} else {
+				require.Nil(t, err)
+				require.Equal(t, success, resp)
+			}
+
+			metadata := mockKafkaClient.recordsPerTopic[".metadata"]
+			require.Equal(t, test.expectedMetadataRecords, len(metadata))
```

Comment:
nit. Do we need to save the mock records in the mock producer for other tests? I guess a counter would be sufficient right?